pkg/retry/retry.go

```diff
@@ -77,3 +81,78 @@ func OnError(
 		}
 	}
 }
+
+// Call executes a function repeatedly until it succeeds or the retry strategy
+// indicates that no more retries should be attempted.
+//
+// If no retry strategy is provided, it defaults to an exponential backoff strategy
+// with 5 max retries, 1000ms initial delay, and 16000ms max delay.
+//
+// Returns the result from the work function and any error that occurred if retries are exhausted.
+func Call[T any](
+	work func() (T, error),
+	retryStrategy ...RetryStrategyFunc,
+) (T, error) {
+	var (
+		result T
+		err    error
+	)
+	if retryStrategy == nil {
+		retryStrategy = []RetryStrategyFunc{WithDefaultExponentialDelay}
+	}
+
+	for retryCount := 0; ; retryCount++ {
+		result, err = work()
+		if err == nil {
+			return result, nil
+		}
+
+		if !retryStrategy[0](retryCount) {
+			return result, err
+		}
+	}
+}
+
+// paramsGetter defines an interface for types that can retrieve parameters of type T.
+// Implementations must provide a GetParams method that accepts a context and returns
+// the retrieved parameters along with any error that occurred during retrieval.
+type paramsGetter[T any] interface {
+	GetParams(ctx context.Context) (T, error)
+}
+
+// GetParams retrieves parameters from a paramsGetter with automatic retry handling.
+//
+// This function wraps the GetParams method of the provided paramsGetter with retry
+// logic using the Call function. If the initial attempt to get parameters fails,
+// subsequent attempts will be made according to the provided retry strategy.
+func GetParams[T any](
+	ctx context.Context,
+	paramsGetter paramsGetter[T],
+	retryStrategy ...RetryStrategyFunc,
+) (T, error) {
+	// Wrap the GetParams call in a function that matches the signature expected by Call
+	// This adapts the paramsGetter.GetParams method to be used with our retry mechanism
+	return Call(
+		func() (T, error) { return paramsGetter.GetParams(ctx) },
+		retryStrategy...,
+	)
+}
+
+// WithExponentialBackoffFn creates a retry strategy with exponential backoff.
```

Comment:
#PUC with an explanation for the case where `retryCount >= maxRetryCount`; i.e. continue retrying with max delay.

Reply:
This seems unresolved. :thinking: